### PR TITLE
Support folding reciprocal ops across `scf.if` region boundary

### PR DESCRIPTION
**Pull request by antiagainst**
This is similar to folding such ops across `scf.for` region boundary.
So place it in the same pass and rename it properly.

### CODE REVIEW

**Comment by antiagainst** [comment]
FYI: there are two commits: the first one implements the logic and the second one is purely renaming. So it might be easier to review each one separately.

File: iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp [hunk 1]
```diff
@@ -50,86 +94,184 @@ namespace {
...
+
+    // Require all yield values in the "else" branch to be splat constant for
+    // now. Merging different cases between the "then" and "else" branches can
+    // be complicated; leave that to only when needed.
+    if (ifOp.elseBlock()) {
+      SplatElementsAttr splat;
+      for (Value v : ifOp.elseYield().getOperands()) {
+        if (!matchPattern(v, m_Constant(&splat))) return failure();
+      }
+    }
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp]: That seems very ad hoc. I think you can make the general case work by just cloning the extract in the else case. This should simplify the logic too.

File: iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp [hunk 2]
```diff
@@ -20,15 +22,57 @@ namespace iree_compiler {
...
+static void transferBody(Block* source, Block* dest, ArrayRef<Value> results,
+                         RewriterBase& rewriter) {
+  // Move all operations to the destination block.
+  rewriter.mergeBlocks(source, dest, dest->getArguments());
+  // Replace the yield op by one that returns only the used values.
+  auto yieldOp = cast<scf::YieldOp>(dest->getTerminator());
+  yieldOp.getOperation()->setOperands(results);
+}
+
+/// Fold reciprocal ops across ForOp boundary. It is common when doing
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp]: The name reciprocal sounds a bit odd to me in this context (makes me think of 1/x kind of reciprocal). Should. we say canceling ops or maybe another name? (If any native speakers are around maybe they have suggestions).
Not a blocker though.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I am not sure having such heavy weight optimization in IREE is a good idea. We should really move this to core.

**Comment by MaheshRavishankar** [comment]
(Also please add me as reviewer to all PRs that are touching `Codegen` layer).

**Comment by ThomasRaoux** [comment]
> I am not sure having such heavy weight optimization in IREE is a good idea. We should really move this to core.

For context, I originally pushed this in IREE as I was getting pushback to have it upstream since it wasn't generic enough:
https://llvm.discourse.group/t/canonicalization-cross-region/2072/9

File: iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp]: Good point! Done. Thanks for the suggestion!

File: iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp]: Yeah I don't have a good term here too.. Changed.

**Comment by antiagainst** [comment]
> I am not sure having such heavy weight optimization in IREE is a good idea. We should really move this to core.

We can give it another push to move to core. What about addressing the comments here and land, then Thomas or I can send a patch upstream and see how folks think.

**Comment by antiagainst** [comment]
Not needed for now. Closing.

**Pull request closed without merging**